### DISCUSSION TITLE
How to use angular limits in physical bones

### DISCUSSION THREAD

**Discussion by sebastianrueckerai** [godot-jolt/godot-jolt#561] [Questions]
Hi!

I am trying to construct an active ragdoll as shown here:
https://www.youtube.com/watch?v=0MHY2TDeMLM&t=3s
...but with Jolt 3D physics enabled (in Godot 4.1.1.stable.mono).

In principle it is working: I can walk around and the bones react to the environment. The problem is that bone constraints seem to be ignored. 

In my example I chose a 6DOF joint for the neck and set the angular limits to 30/-30 for each axis.
As can be seen in the screenshot, the neck bends way more than that.

Do I need to do anything special to get this to work? As I understand it, activating Jolt physics should make this work without doing anything else as Godot will be using the jolt implementation of these joints there limits and springs work, right?

I would be super thankful for any help here. I can also share my (almost) working versiopn of the active ragdoll scene, if that helps.

![image](https://github.com/godot-jolt/godot-jolt/assets/83586913/16480e84-a42f-446b-82c7-630b3ce5e422)
![image](https://github.com/godot-jolt/godot-jolt/assets/83586913/c87597f9-f365-4939-9e14-0ebe80576884)

**Comment by mihe** [comment]
Judging by that last screenshot it looks like the transform of your neck bone is quite tilted, which I believe will affect the joint transform, which in turn is what acts as the basis/origin for the joint limits.

**Reply by sebastianrueckerai** [reply]
Hey! :)
Can you elaborate what you mean by that?
I attached two pictures of the rig which show the relative positions of the bones.

I cant quite see why the relative bone positions should change anything though. 
The origin of each bone around which a joint moves should be the tip of it's parent bone. 
The axis of this joint are given by the bones local coordinate system. 
All this seems to be reflected accurately by the widget showing the limits of the 6DoF joint. 
But maybe I am not getting your point.

![image](https://github.com/godot-jolt/godot-jolt/assets/83586913/6da4f14d-7fd4-4533-a7b3-c6a6b5d2d651)
![image](https://github.com/godot-jolt/godot-jolt/assets/83586913/f2e739d1-84f7-478e-8684-9a9a6bc4f8ae)

**Reply by mihe** [reply]
I seem to have misunderstood your problem, now that I'm reading it again, and I'm not sure my answer made much sense even for the problem I thought you were having.

Upload the project/scene and I'll take a look.

**Reply by sebastianrueckerai** [reply]
Gladly, thanks!
Here you go:
https://www.dropbox.com/scl/fo/ej6avj2vri6uqxnas1act/h?rlkey=pmcluihr8t8np6caa5sgae5n2&dl=0

The "ghost" is the slightly transparent model that shows just the animations the model has.
The solid model is a physical skeleton that is being "pushed" into a position closely resembling that of the animated skeleton.
Since the solid model is interacting with the environment it can only follow the animations if no obstacle is in the way.

The code for that comes from "Ragdoll.gd" and is taken from the youtube video linked above.

**Reply by mihe** [reply]
I've been staring at this for a bit now, and I must admit that this type of stuff is not my strong suit, so I don't have any good answers for you, unfortunately.

Without having actually looked at any numbers, I get the sense that the limits are working, but that whatever impulses are being applied by the joints themselves (in order to keep the bodies at their limits) are being overwhelmed by the velocities applied in `Ragdoll.gd`. I'm mostly just guessing though, and that's not much of an educated guess either.

Seeing as how this happens with Godot Physics as well I'm assuming there's some fundamental problem here rather than a Jolt-specific one.

I did try the original project from that YouTube video as well, and while the limits seem to actually do something there (when shooting the bowling balls at it) Jolt does not seem to work well with the way that's set up. I believe that has to do with the fact that in this extension the default joints are all unsprung hard limit joints, which seem to freak out a bit when connected in series, especially with a zero limit like what's found in that project.

Jolt does technically have some kind of pose-matching as part of its own ragdoll support, as seen [here](https://youtu.be/lYHhe6HLbs4), which seem to use "positional motors" (springs) coupled with target transforms for its bone joints instead, which clearly seems to work well. While I do support some of that stuff, I don't expose the ragdoll stuff in any way, and likely won't for the foreseeable future, unfortunately.

**Reply by sebastianrueckerai** [reply]
Thanks for taking the time! 🙏 
Can I sponsor you to work on exposing jolts pose matching?
I could also help, though this area would be new to me.
I am "sebas714n" on discord, in case you are active there!

**Reply by mihe** [reply]
> Can I sponsor you to work on exposing jolts pose matching?

I appreciate the offer, but that's not something I'm interested in currently.

> I could also help, though this area would be new to me.

I'm spread pretty thin right now, and about to be even more so in the coming weeks, so whoever integrates this would need to be fairly autonomous. At the same time I would prefer to be involved in how this gets integrated, so it would still eat into time I don't really have.

There's also the issue of the C# scripting not supporting nodes exposed from GDExtension. I've tried to not make C# a second-class citizen, by focusing mainly on the drop-in aspect of this extension, and exposing Jolt-specific nodes undermines that a bit.

**Reply by sebastianrueckerai** [reply]
I understand, so many things to do, so little time...

In the meantime I found out that if I dial down the force which pushes the bones into place , the limits seem to work again.
So it seems like your theory of these forces being too strong in comparison to those which guard the limits may be true.
Unfortunately, with low forces on the pose matching, the whole character becomes a sluggish zombie (though one with working joint limits).

I will try and play with the stiffness of the joints. This did not work in pure Godot but maybe it will increase the forces that make the joint limits happen in jolt.
